cmd/juju/charmhub/printinfo.go

```diff
@@ -0,0 +1,535 @@
+	if err := c.maybePrintStoreURL(); err != nil {
+		return errors.Trace(err)
+	}
+	if err := c.maybePrintDescription(); err != nil {
+		return errors.Trace(err)
+	}
+	if err := c.maybePrintResouces(); err != nil {
+		return errors.Trace(err)
+	}
+	if err := c.maybePrintRelations(); err != nil {
+		return errors.Trace(err)
+	}
+	if err := c.maybePrintChannels(); err != nil {
+		return errors.Trace(err)
+	}
+	return nil
+}
+
+func (c *charmInfoWriter) maybePrintID() error {
+	if c.in.ID == "" {
+		return nil
+	}
+
+	_, err := fmt.Fprintf(c.w, "charm-id:    %s\n", c.in.ID)
+	return err
+}
+
+func (c *charmInfoWriter) maybePrintSupports() error {
+	if c.charmMeta == nil {
+		return nil
+	}
+	if len(c.charmMeta.Series) == 0 {
+		return nil
+	}
+	_, err := fmt.Fprintf(c.w, "supports:    %s\n", strings.Join(c.charmMeta.Series, ", "))
+	return err
+}
+
+func (c *charmInfoWriter) maybePrintTags() error {
+	if c.charmMeta == nil {
+		return nil
+	}
+	if len(c.charmMeta.Tags) == 0 {
+		return nil
+	}
+	_, err := fmt.Fprintf(c.w, "tags:        %s\n", strings.Join(c.charmMeta.Tags, ", "))
+	return err
+}
+
+func (c *charmInfoWriter) maybePrintSubordinate() error {
+	if c.charmMeta == nil {
+		return nil
+	}
+	_, err := fmt.Fprintf(c.w, "subordinate: %t\n", c.charmMeta.Subordinate)
+	return err
+}
+
+func (c *charmInfoWriter) maybePrintResouces() error {
+	if c.charmMeta == nil {
+		return nil
+	}
+	if len(c.charmMeta.Resources) == 0 {
+		return nil
+	}
+	if _, err := fmt.Fprintln(c.w, "resources:"); err != nil {
+		return err
+	}
+	for _, meta := range c.charmMeta.Resources {
+		_, err := fmt.Fprintf(c.w, "  %s (%s)\n", meta.Name, meta.Type.String())
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (c *charmInfoWriter) maybePrintRelations() error {
+	if c.charmMeta == nil {
+		return nil
+	}
+	if len(c.charmMeta.Requires) == 0 && len(c.charmMeta.Provides) == 0 {
+		return nil
+	}
+	if _, err := fmt.Fprintf(c.w, "relations:"); err != nil {
+		return err
+	}
+	if provides, ok := formatRelationPart(c.charmMeta.Provides); ok {
+		if _, err := fmt.Fprintf(c.w, "  provides:\n%s", provides); err != nil {
+			return err
+		}
+	}
+	if requires, ok := formatRelationPart(c.charmMeta.Requires); ok {
+		if _, err := fmt.Fprintf(c.w, "  requires:\n%s", requires); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func formatRelationPart(rels map[string]charm.Relation) (string, bool) {
+	if len(rels) <= 0 {
+		return "", false
+	}
+	var relations string
+	for k, v := range rels {
+		relations = relations + "  " + k + ": " + v.Name + "\n"
+	}
+	return relations, true
+}
+
+func (c *charmInfoWriter) maybeUnmarshalCharmMetadata() {
+	if c.in.DefaultRelease.Revision.MetadataYAML == "" {
+		return
+	}
+	m := c.in.DefaultRelease.Revision.MetadataYAML
+	meta, err := charm.ReadMeta(bytes.NewBufferString(m))
+	if err != nil {
+		// Do not fail on unmarshalling metadata, log instead
+		c.warningf(errors.Annotate(err, "cannot unmarshal charm metadata").Error())
```

Comment:
Why? This should always be valid?

Reply:
yes, should always be valid.  however our live test data is not.  :-/